src/headers.jl

```diff
@@ -0,0 +1,246 @@
+# See https://httpwg.github.io/specs/rfc7230.html#rule.token.separators
+const validHeaderBytes = IntSet(Int(c) for c in
+   raw"!#$%&'*+-.0123456789ABCDEFGHIJKLMNOPQRSTUWVXYZ^_`abcdefghijklmnopqrstuvwxyz|~"
+)
+
+validHeaderChar(c::Char) = isascii(c) && UInt8(c) in validHeaderBytes
+
+function canonicalHeaderKey(s::String)
+    # fast path for valid header keys
+    upper = true
+    for c in s
+        if !validHeaderChar(c)
+            return s
+        end
+        if upper && 'a' <= c <= 'z'
+            return _chk(s)
+        end
+        if !upper && 'A' <= c <= 'Z'
+            return _chk(s)
+        end
+
+        upper = c == '-'
+    end
+
+    return s
+end
+
+const toUpper = 'A' - 'a'
+
+function _chk(s::String)
+    # confirm its fixable before changing anything
+    for c in s
+        if !validHeaderChar(c)
+            return s
+        end
+    end
+
+    a = copy(Array{UInt8}(s))
+    upper = true
+    for (i, b) in enumerate(a)
+        if upper && UInt8('a') <= b <= UInt8('z')
+            b += toUpper
+        elseif !upper && UInt8('A') <= b <= UInt8('Z')
+            b -= toUpper
+        end
+        @inbounds a[i] = b
+        upper = b == UInt8('-')
+    end
+
+    return String(a)
+end
+
+"""
+Internal type used to enable fast merge, copy, and filter of Headers
+"""
+immutable AlreadyCanonical
```

Comment:
I think having the boolean flag might also be nice because it provides an easier "escape hatch", if there are cases where the user needs to provide a header as-is (not exactly sure on the use-case here, but some kind of custom setup where the server only accepts specific header keys). It's a little easier to do `Header(d, alreadycanonical=true)` than fiddling w/ an HTTP internal type.